rand_distr/src/gamma.rs

```diff
@@ -542,31 +572,144 @@ impl std::error::Error for BetaError {}
     Open01: Distribution<F>,
 {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
-        let x = self.gamma_a.sample(rng);
-        let y = self.gamma_b.sample(rng);
-        x / (x + y)
+        match self.algorithm {
+            BetaAlgorithm::BB(algo) => {
+                let mut w;
+                loop {
+                    // 1.
+                    let u1 = rng.sample(Open01);
+                    let u2 = rng.sample(Open01);
+                    let v = algo.beta * (u1 / (F::one() - u1)).ln();
+                    w = self.a * v.exp();
+                    let z = u1*u1 * u2;
+                    let r = algo.gamma * v - F::from(4.).unwrap().ln();
+                    let s = self.a + r - w;
+                    // 2.
+                    if s + F::one() + F::from(5.).unwrap().ln()
+                        >= F::from(5.).unwrap() * z {
+                        break;
+                    }
+                    // 3.
+                    let t = z.ln();
+                    if s >= t {
+                        break;
+                    }
+                    // 4.
+                    if !(r + algo.alpha * (algo.alpha / (self.b + w)).ln() < t) {
+                        break;
+                    }
+                }
+                // 5.
+                if !self.switched_params {
+                    w / (self.b + w)
+                } else {
+                    self.b / (self.b + w)
+                }
+            },
+            BetaAlgorithm::BC(algo) => {
+                let mut w;
+                loop {
+                    let z;
+                    // 1.
+                    let u1 = rng.sample(Open01);
+                    let u2 = rng.sample(Open01);
+                    if u1 < F::from(0.5).unwrap() {
+                        // 2.
+                        let y = u1 * u2;
+                        z = u1 * y;
+                        if F::from(0.25).unwrap() * u2 + z - y >= algo.kappa1 {
+                            continue;
+                        }
+                    } else {
+                        // 3.
+                        z = u1 * u1 * u2;
+                        if z <= F::from(0.25).unwrap() {
+                            let v = algo.beta * (u1 / (F::one() - u1)).ln();
+                            w = self.a * v.exp();
+                            break;
+                        }
+                        // 4.
+                        if z >= algo.kappa2 {
+                            continue;
+                        }
+                    }
+                    // 5.
+                    let v = algo.beta * (u1 / (F::one() - u1)).ln();
+                    w = self.a * v.exp();
+                    if !(algo.alpha * ((algo.alpha / (self.b + w)).ln() + v)
+                         - F::from(4.).unwrap().ln() < z.ln()) {
+                        break;
+                    };
+                }
+                // 6.
+                if !self.switched_params {
```

Comment:
Now, this condition is a bit tricky because you decided to set `a = min(a0, b0)` and `b = max(a0, b0)` for both algorithms BB and BC.

From the paper, the switching condition can be read as follows:
1) $W / (b + W)$ follows a distribution Beta(max(a0, b0), min(a0, b0))
2) $b / (b + W)$ follows a distribution Beta(min(a0, b0), max(a0, b0)).

Therefore, the conditions are the other way around!! 
Following your code:
`switched_params = false` only if `a0 = min(a0, b0)`.
Therefore, if `switched_params = false`, then the correct variable is $b / (b + W)$, not $W / (b + W)$.

Reply:
I'll use the same `a` and `b` as in the paper instead, then this should be fine.